File: PHP/src/OneLang/One/Transforms/InferTypesPlugins/NullabilityCheckWithNot.php
<?php

namespace OneLang\One\Transforms\InferTypesPlugins\NullabilityCheckWithNot;

use OneLang\One\Transforms\InferTypesPlugins\Helpers\InferTypesPlugin\InferTypesPlugin;
use OneLang\One\Ast\Expressions\Expression;
use OneLang\One\Ast\Expressions\UnaryExpression;
use OneLang\One\Ast\Expressions\BinaryExpression;
use OneLang\One\Ast\Expressions\NullLiteral;
use OneLang\One\Ast\AstTypes\ClassType;

class NullabilityCheckWithNot extends InferTypesPlugin {
    function __construct() {
        parent::__construct("NullabilityCheckWithNot");
        
    }
    
    function canTransform($expr) {
        return $expr instanceof UnaryExpression ? $expr->operator === "!" : false;
    }
    
    function transform($expr) {
        $unaryExpr = $expr;
        if ($unaryExpr->operator === "!") {
            $this->main->processExpression($expr);
            $type = $unaryExpr->operand->actualType;
            $litTypes = $this->main->currentFile->literalTypes;
            if ($type instanceof ClassType && $type->decl !== $litTypes->boolean->decl && $type->decl !== $litTypes->numeric->decl)
                return new BinaryExpression($unaryExpr->operand, "==", new NullLiteral());
        }
        
        return $expr;
    }
}

File: Python/OneLang/One/Transforms/InferTypesPlugins/NullabilityCheckWithNot.py
from onelang_core import *
import OneLang.One.Transforms.InferTypesPlugins.Helpers.InferTypesPlugin as inferTypesPlug
import OneLang.One.Ast.Expressions as exprs
import OneLang.One.Ast.AstTypes as astTypes

class NullabilityCheckWithNot(inferTypesPlug.InferTypesPlugin):
    def __init__(self):
        super().__init__("NullabilityCheckWithNot")
    
    def can_transform(self, expr):
        return expr.operator == "!" if isinstance(expr, exprs.UnaryExpression) else False
    
    def transform(self, expr):
        unary_expr = expr
        if unary_expr.operator == "!":
            self.main.process_expression(expr)
            type = unary_expr.operand.actual_type
            lit_types = self.main.current_file.literal_types
            if isinstance(type, astTypes.ClassType) and type.decl != lit_types.boolean.decl and type.decl != lit_types.numeric.decl:
                return exprs.BinaryExpression(unary_expr.operand, "==", exprs.NullLiteral())
        
        return expr